### PR TITLE
[Bifrost] LogletOffset is now u32

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[Bifrost] LogletOffset is now u32

This type fits better the model of the virtual log where the entire LSN address space is u64.

Bonus: I noticed an occasional segfault in partition tests due to improper rocksdb shutdown, so I fixed it.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1880).
* #1883
* __->__ #1880
* #1879

### CODE REVIEW

**Review by pcholakov — APPROVED** [review]

File: crates/bifrost/src/loglet/mod.rs [hunk 1]
```diff
@@ -48,18 +48,34 @@ use crate::{Result, TailState};
...
+    }
+}
+
+impl Add<u32> for LogletOffset {
     type Output = Self;
-    fn add(self, rhs: usize) -> Self {
-        // we always assume that we are running on a 64bit cpu arch.
-        Self(self.0.saturating_add(rhs as u64))
+    fn add(self, rhs: u32) -> Self {
+        // checked add (in debug mode)
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/loglet/mod.rs:68]: Whoops. Correct. I've changed the code from + to checked_add and didn't update the comment (as per usual). Will update.

File: crates/bifrost/src/loglet/mod.rs [hunk 2]
```diff
@@ -48,18 +48,33 @@ use crate::{Result, TailState};
...
-impl Add<usize> for LogletOffset {
+impl LogletOffset {
+    pub const fn new(offset: u32) -> Self {
+        Self(offset)
+    }
+}
+
+impl From<LogletOffset> for u64 {
+    fn from(value: LogletOffset) -> Self {
+        value.0 as u64
```
Comment (tillrohrmann) [inline crates/bifrost/src/loglet/mod.rs:61]: `u64::from(value.0)`?

File: crates/bifrost/src/providers/local_loglet/keys.rs [hunk 3]
```diff
@@ -49,6 +49,8 @@ impl RecordKey {
         buf.reserve(Self::serialized_size());
         buf.put_u8(b'd');
         buf.put_u64(self.loglet_id);
+        // despite offset being 32bit, for backward compatibility and future-proofing, we keep the
+        // storage layer as u64 but we assert that the actual value fits u32 on read.
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/local_loglet/keys.rs:53]: 👍

File: crates/bifrost/src/providers/local_loglet/log_store_writer.rs [hunk 4]
```diff
@@ -325,7 +325,9 @@ impl RocksDbLogWriterHandle {
         payloads: Arc<[ErasedInputRecord]>,
     ) -> Result<AckRecv, ShutdownError> {
         let (ack, receiver) = oneshot::channel();
-        let last_offset_in_batch = (start_offset + payloads.len()).prev();
+        // Do not allow more than 65k records in a single batch!
+        assert!(payloads.len() < u16::MAX as usize);
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/local_loglet/log_store_writer.rs:329]: `<=`?

File: crates/bifrost/src/providers/local_loglet/mod.rs [hunk 5]
```diff
@@ -152,7 +152,10 @@ impl Loglet for LocalLoglet {
             return Ok(LogletCommit::sealed());
         }
 
-        let num_payloads = payloads.len();
+        // Do not allow more than 65k records in a single batch!
+        assert!(payloads.len() < u16::MAX as usize);
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/local_loglet/mod.rs:156]: Is it deliberate to exclude `u16::MAX`?

File: crates/bifrost/src/providers/local_loglet/read_stream.rs [hunk 6]
```diff
@@ -232,7 +232,7 @@ impl Stream for LocalLogletReadStream {
                 // We have a bug! we shouldn't be in this location where the record
                 // doesn't exist but we expect it to!
                 error!(
-                    log_id = self.log_id,
+                    log_id = self.loglet_id,
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/local_loglet/read_stream.rs:235]: `loglet_id = self.loglet_id`?

File: crates/worker/src/partition/state_machine/tests/mod.rs [hunk 7]
```diff
@@ -183,6 +189,7 @@ async fn start_invocation() -> TestResult {
 
     let invocation_status = test_env.storage().get_invocation_status(&id).await.unwrap();
     assert_that!(invocation_status, pat!(InvocationStatus::Invoked(_)));
+    test_env.shutdown().await;
```
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/tests/mod.rs:192]: 🙏

**Review by tillrohrmann — APPROVED** [review]
Nice work @AhmedSoliman. LGTM. +1 for merging.

File: crates/bifrost/src/providers/local_loglet/mod.rs (hunk 5 above)
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/local_loglet/mod.rs:156]: No. Will fix.

**Pull request merged**